atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -709,8 +709,8 @@ private void extractLocalWritesForRow(@Output Map<Cell, byte[]> result,
                         .map(rangeRequest -> Pair.of(rangeRequest, getLazyRange(tableRef, rangeRequest)))
                         .collect(Collectors.toList());
 
-        return ConcurrentStreams.map(
-                requestAndVisitables,
+        return MoreStreams.blockingStreamWithParallelism(
+                requestAndVisitables.stream(),
```

Comment:
Oo now that this takes a stream, can we just remove the collect above?

Reply:
Also may want to handle the case where concurrency or size is 1 to avoid the scheduling overhead (will become useful as we start to adjust concurrency as a function of load). So maybe something like:
```
        Stream<Pair<RangeRequest, BatchingVisitable<RowResult<byte[]>>>> requestAndVisitables =
                StreamSupport.stream(rangeRequests.spliterator(), false)
                        .map(rangeRequest -> Pair.of(rangeRequest, getLazyRange(tableRef, rangeRequest)));

        if (concurrencyLevel == 1 || Iterables.size(rangeRequests) == 1) {
            return requestAndVisitables.map(pair -> visitableProcessor.apply(pair.getLeft(), pair.getRight());
        }

        return MoreStreams.blockingStreamWithParallelism(
                requestAndVisitables,
                pair -> visitableProcessor.apply(pair.getLeft(), pair.getRight()),
                getRangesExecutor,
                concurrencyLevel);
```

Reply:
actually, for now I'm cool with this short circuit. However, I don't want to do the Iterables.size one - if we want to leverage the 1 case I think it should be a Collection we take so we know we can get the size efficiently - otherwise it's a confusing API.

Reply:
(or I can carry on collecting - what do you think?)

Reply:
By efficiently you mean not iterating through it in linear time? I would guess all our use-cases give it a `List`, which `Iterables#size` should do in constant time. But understood if that's weird from a general-case/API perspective.

Reply:
If not, though, I think leaving the size check out and not collecting would be better since common case will be > 1 and that'll reduce memory usage.

Reply:
If we want to special case 0 or 1 ranges, could just check the iterable so its still O(2):

```
        if (Iterables.isEmpty(elements) || isSingleton(elements)) {
            return StreamSupport.stream(elements.spliterator(), false);
        }
        // parallelize

    private static <T> boolean isSingleton(Iterable<T> elements) {
        Iterator<T> it = elements.iterator();
        if (it.hasNext()) {
            it.next();
            return !it.hasNext();
        }
        return false;
    }


    @Test
    public void singletonStream() throws Exception {
        assertThat(isSingleton(ImmutableList.of("a"))).isTrue();
        assertThat(isSingleton(ImmutableList.of())).isFalse();
        assertThat(isSingleton(ImmutableList.of("a", "b"))).isFalse();
    }
```

Reply:
yep, cool.